### ISSUE TITLE
Implement testing framework and fix bug in 3.5

### ISSUE THREAD

**Issue by Jbwasse2** [tauhid03/MarkovChainAnalysis#14]

**Comment by Jbwasse2** [comment]
The testing framework is through github to check different versions of python that this library will support.

**Comment by Jbwasse2** [comment]
The bug only occurs in test_kron_inv's test_random_pinv function. For some reason it also only happens when p=n=5. I have tried it for many other cases such as n=p=4, or n=3 and p=16. But this error could not be created again with an easier to debug dataset...

**Comment by alexandroid000** [comment]
Error replicated on Arch Linux with Python 3.7.2, numpy 1.16.2, p=n=5 and p=n=6, but passes on p=n=4.

In case it was a pass-by-reference issue, I tried replacing

```
for A in r_As:
    A[1,:] = A[0,:]
```

with

```
for A in r_As:
    A[1,:] = deepcopy(A[0,:])
```

to no effect.

My only hunch is it's a propagating floating point error, and when the matrices exceed a certain size it causes problems? Maybe a contributing factor is that the matrix inverse is underdetermined, maybe combined with some bug in the `ortho_group` class? But then why wouldn't it also fail on `testBig_pInv`?

**Comment by Jbwasse2** [comment]
https://travis-ci.com/Jbwasse2/MarkovChainAnalysis

It passed the test in 3.5 and 3.6 here.

**Comment by Jbwasse2** [comment]
Here is an error file I created, I haven't looked at it yet. I had it print out information about the solver. It would run until it hit a fail case (which took 19 iterations).

[debug.txt](https://github.com/tauhid03/MarkovChainAnalysis/files/2981407/debug.txt)

**Comment by Jbwasse2** [comment]
But I think you are right, it doesn't make sense for it to not fail on the smaller matrices, and only on the bigger ones.

**Comment by Jbwasse2** [comment]
Here is a n=4,p=4 case of it failing

[debug.txt](https://github.com/tauhid03/MarkovChainAnalysis/files/2981529/debug.txt)

**Comment by Jbwasse2** [comment]
I think I figured it out. In the kronprodinv we would pinv the matrices and then do the dot product. In the full case we get the kron product of the matrices then do pinv. It seems like there are cases where this causes problems. Although in fairness this shouldn't be an issue since $(A_1 \otimes A_2)^* = A_1^* \otimes A_2^*$ according to [wikipedia](https://en.wikipedia.org/wiki/Kronecker_product#Relations_to_other_matrix_operations). Although in fairness it says it needs to be invertible for that property to hold, but then why mention pinv at all?

**Comment by Jbwasse2** [comment]
This did fix the build problem. https://travis-ci.com/Jbwasse2/MarkovChainAnalysis

**Comment by Jbwasse2** [comment]
Fixed in #15
